### PR TITLE
Adding registration tests

### PR DESCRIPTION
**Pull request by PavelApis**

### CODE REVIEW

File: src/test/java/io/github/enkarin/bookcrossing/service/UserServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    @Test
+    void findAllTest() {
+        final List<UserDto> users = TestDataProvider.buildUsers().stream()
+                .map(u -> userService.saveUser(u))
+                .collect(Collectors.toList());
+        users.forEach(u -> usersId.add(u.getUserId()));
+
+        final var foundUsers = userService.findAllUsers(0);
```
Comment (EnKarin) [inline src/test/java/io/github/enkarin/bookcrossing/service/UserServiceTest.java:110]: Admin functions have already been tested in AdminTest

File: src/test/java/io/github/enkarin/bookcrossing/support/TestDataProvider.java [hunk 2]
```diff
@@ -125,4 +125,24 @@ public static UserRegistrationDto buildAlex() {
                 .email("t.test@mail.ru")
                 .build();
     }
+
+    @Nonnull
+    public static UserRegistrationDto buildNonConfirmedPasswordUser() {
+        return UserRegistrationDto.builder()
```
Comment (EnKarin) [inline src/test/java/io/github/enkarin/bookcrossing/support/TestDataProvider.java:131]: You can use the prepareUser() to not write the definition of some fields

File: src/test/java/io/github/enkarin/bookcrossing/service/UserServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+
+public class UserServiceTest extends BookCrossingBaseTests {
+    @Autowired
+    MailService mailService;
+
+    @AfterEach
+    void deleteAddedUsers() {
```
Comment (EnKarin) [inline src/test/java/io/github/enkarin/bookcrossing/service/UserServiceTest.java:27]: There is in the base class

File: src/test/java/io/github/enkarin/bookcrossing/service/UserServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        final UserDto userDTO = userService.saveUser(TestDataProvider.buildAlex());
+        usersId.add(userDTO.getUserId());
+
+        final UserProfileDto userProfileDto = userService.getProfile(userDTO.getLogin());
+        checkUserDTOAndUserProfileDTO(userDTO, userProfileDto);
+    }
+
+    @Test
+    void findAllIsEmptyListTest() {
+        assertThat(userService.findAllUsers(0)).isEmpty();
```
Comment (EnKarin) [inline src/test/java/io/github/enkarin/bookcrossing/service/UserServiceTest.java:100]: Already tested

File: src/test/java/io/github/enkarin/bookcrossing/service/UserServiceTest.java (hunk 1 above)
Comment (PavelApis) [reply src/test/java/io/github/enkarin/bookcrossing/service/UserServiceTest.java:110]: But these are different methods in project logic, maybe I missed something

I think it is better to test them separately or we need to reuse the findAll() from the UserService in AdminService

File: src/test/java/io/github/enkarin/bookcrossing/support/TestDataProvider.java (hunk 2 above)
Comment (PavelApis) [reply src/test/java/io/github/enkarin/bookcrossing/support/TestDataProvider.java]: Sure) Changed it

File: src/test/java/io/github/enkarin/bookcrossing/service/UserServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+
+public class UserServiceTest extends BookCrossingBaseTests {
```
Comment (mfvanek) [inline src/test/java/io/github/enkarin/bookcrossing/service/UserServiceTest.java:19]: `public` modifier is not required here

File: src/test/java/io/github/enkarin/bookcrossing/service/UserServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    @Test
+    void saveUserCorrectUserTest() {
+        final UserDto user = userService.saveUser(TestDataProvider.buildAlex());
+        usersId.add(user.getUserId());
```
Comment (mfvanek) [inline src/test/java/io/github/enkarin/bookcrossing/service/UserServiceTest.java:23]: Try to add method like `createUser(...) `and move this logic `usersId.add(user.getUserId());` there

File: src/test/java/io/github/enkarin/bookcrossing/service/UserServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+        assertThat(userService.findById(user.getUserId(), 0)).isNotNull();
```
Comment (mfvanek) [inline src/test/java/io/github/enkarin/bookcrossing/service/UserServiceTest.java:25]: What does zero mean here?

**Review by mfvanek — CHANGES_REQUESTED** [review]

File: src/test/java/io/github/enkarin/bookcrossing/service/UserServiceTest.java (hunk 7 above)
Comment (PavelApis) [reply src/test/java/io/github/enkarin/bookcrossing/service/UserServiceTest.java]: UTC±0:00 time zone
Added special constant in class

File: src/test/java/io/github/enkarin/bookcrossing/service/UserServiceTest.java (hunk 5 above)
Comment (PavelApis) [reply src/test/java/io/github/enkarin/bookcrossing/service/UserServiceTest.java]: Done

File: src/test/java/io/github/enkarin/bookcrossing/service/UserServiceTest.java (hunk 6 above)
Comment (mfvanek) [reply src/test/java/io/github/enkarin/bookcrossing/service/UserServiceTest.java:24]: Now let's move createUser to the base class

**Review by mfvanek — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 8]
```diff
@@ -142,6 +142,35 @@
...
+                        <configuration>
+                            <rules>
+                                <bannedPlugins>
+                                    <level>ERROR</level>
+                                    <excludes>
+                                        <exclude>org.apache.maven.plugins:maven-verifier-plugin</exclude>
+                                    </excludes>
+                                </bannedPlugins>
+                                <requireMavenVersion>
+                                    <version>3.8.3</version>
```
Comment (mfvanek) [inline pom.xml:164]: https://maven.apache.org/pom.html#dependency-version-requirement-specification
Add ability to use with more recent versions

File: pom.xml [hunk 9]
```diff
@@ -153,6 +153,35 @@
...
+                <artifactId>maven-enforcer-plugin</artifactId>
+                <version>3.1.0</version>
+                <executions>
+                    <execution>
+                        <id>enforce-versions</id>
+                        <goals>
+                            <goal>enforce</goal>
+                        </goals>
+                        <configuration>
+                            <rules>
```
Comment (mfvanek) [inline pom.xml:167]: add `<banDuplicatePomDependencyVersions />` rule

File: pom.xml (hunk 8 above)
Comment (mfvanek) [reply pom.xml:175]: for example
```
                                <requireMavenVersion>
                                    <version>[3.6,)</version>
                                </requireMavenVersion>
                                <requireJavaVersion>
                                    <version>[11,)</version>
                                </requireJavaVersion>
```

File: src/test/java/io/github/enkarin/bookcrossing/base/BookCrossingBaseTests.java [hunk 10]
```diff
@@ -38,6 +40,12 @@ public abstract class BookCrossingBaseTests {
             .withConfiguration(GreenMailConfiguration.aConfig().withDisabledAuthentication())
             .withPerMethodLifecycle(false);
 
+    protected UserDto createAndSaveUser() {
+        final UserDto user = userService.saveUser(TestDataProvider.buildAlex());
```
Comment (mfvanek) [inline src/test/java/io/github/enkarin/bookcrossing/base/BookCrossingBaseTests.java:44]: move `TestDataProvider.buildAlex()` to method arguments

**Review by mfvanek — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 11]
```diff
@@ -153,6 +153,30 @@
 
     <build>
         <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-enforcer-plugin</artifactId>
+                <version>3.1.0</version>
```
Comment (mfvanek) [inline pom.xml:159]: In Spring pom <maven-enforcer-plugin.version>3.0.0</maven-enforcer-plugin.version>

**Review by mfvanek — CHANGES_REQUESTED** [review]

File: src/test/java/io/github/enkarin/bookcrossing/base/BookCrossingBaseTests.java [hunk 12]
```diff
@@ -38,6 +40,12 @@ public abstract class BookCrossingBaseTests {
             .withConfiguration(GreenMailConfiguration.aConfig().withDisabledAuthentication())
             .withPerMethodLifecycle(false);
 
+    protected UserDto createAndSaveUser(final UserRegistrationDto userRegistrationDto) {
+        final UserDto user = userService.saveUser(userRegistrationDto);
+        usersId.add(user.getUserId());
```
Comment (mfvanek) [inline src/test/java/io/github/enkarin/bookcrossing/base/BookCrossingBaseTests.java:45]: Now make `usersId` **private** in the base class

File: src/test/java/io/github/enkarin/bookcrossing/registation/controllers/RegistrationControllerTest.java [hunk 13]
```diff
@@ -31,7 +31,7 @@ void registerUserTest() throws MessagingException {
                 .returnResult().getResponseBody();
         assertThat(user).isNotNull();
 
-        usersId.add(user.getUserId());
+        super.trackUserId(user.getUserId());
```
Comment (mfvanek) [inline src/test/java/io/github/enkarin/bookcrossing/registation/controllers/RegistrationControllerTest.java:34]: Remove unnecessary  `super.`

**Review by mfvanek — DISMISSED** [review]

**Review by mfvanek — APPROVED** [review]

**Pull request merged**